autogluon/utils/tabular/ml/models/tab_transformer/TabTransformer_model.py

```diff
@@ -0,0 +1,320 @@
+                    pretext=pretext_tasks['SUPERVISED_pretext'](self.kwargs)
+                elif state=='pretrain':
+                    optimizers = [optim.Adam(self.parameters(), lr=lr, weight_decay=weight_decay)]
+                    pretext=pretext_tasks['BERT_pretext'](self.kwargs)
+                elif state=='finetune':
+                    base_exp_decay=self.kwargs['tab_kwargs']['base_exp_decay']
+                    optimizer_fc    = optim.Adam(self.fc.parameters(), lr=lr, weight_decay=weight_decay) 
+                    optimizer_embeds = optim.Adam(self.embed.parameters(), lr=lr, weight_decay=weight_decay) 
+                    scheduler = optim.lr_scheduler.ExponentialLR(optimizer_embeds,gamma=base_exp_decay)
+                    optimizers.append(optimizer_fc)
+                    optimizers.append(optimizer_embeds)
+
+                    pretext=pretext_tasks['SUPERVISED_pretext'](self.kwargs)
+
+                else:
+                    raise NotImplementedError("state must be one of [None, 'pretrain', 'finetune']")
+
+                if self.kwargs['problem_type']=='regression':
+                    loss_criterion = nn.MSELoss()
+                else:
+                    loss_criterion = nn.CrossEntropyLoss()
+
+                old_val_accuracy=0.0
+         
+                epochs = self.kwargs['pretrain_epochs'] if state=='pretrain' else self.kwargs['epochs']
+                freq   = self.kwargs['pretrain_freq'] if state=='pretrain' else self.kwargs['freq']
+   
+                for e in range(1,epochs+1):
+                    _ = utils.epoch(self, trainloader, optimizers, loss_criterion=loss_criterion, \
+                                        pretext=pretext, state=state, scheduler=None, epoch=e, epochs=epochs, aug_kwargs=self.kwargs['augmentation']) #returns train_loss, train_acc@1, train_acc@5 
+                    
+                    if valloader is not None:
+                        if e % freq == 0:
+                            _, val_accuracy = utils.epoch(self, valloader, optimizers=None, \
+                                loss_criterion=loss_criterion, pretext=pretext, state=state, scheduler=None, epoch=1, epochs=1, aug_kwargs=self.kwargs['augmentation'])                       
+                            if val_accuracy>old_val_accuracy:
+                                torch.save(self,'tab_trans_temp.pth')
+                if valloader is not None:
+                    try:
+                        self=torch.load('tab_trans_temp.pth')
+                        os.remove('tab_trans_temp.pth')
+                    except:
+                        pass        
+
+            
+class TabTransformerModel(AbstractModel):
+    params_file_name="tab_trans_params.pth"
+    def __init__(self,**kwargs):
+        super().__init__(**kwargs)
+        self.types_of_features=None
+
+
+    def _get_types_of_features(self, df):
```

Comment:
This is duplicated code.  Instead, you can move the functions you need out fo the TabularNeuralNetwork model to be accessible by both the TabTransformer and the TabularNeuralNet.